v2/crowdnotifier.ts

```diff
@@ -0,0 +1,220 @@
+import {
+    compare,
+    crypto_box_keypair,
+    crypto_box_seal,
+    crypto_box_seal_open,
+    crypto_hash_sha256,
+    IKeyPair,
+    randombytes_buf
+} from "../lib/sodium";
+import {IEncryptedData ,IBEPrimitives, NONCE_LENGTH} from "./crypto";
+import * as mcl from "../lib/mcl";
+import {
+    IBEIdInternal1,
+    IBEIdInternal2
+} from "./proto";
+
+
+
+export interface IMasterTrace {
+    mpk: mcl.G2;
+    mskv: mcl.Fr;
+    mskhEnc: Uint8Array; // In the paper ctxtHA
+    info: string;
+    nonce1: Uint8Array;
+    nonce2: Uint8Array;
+}
+
+export interface IEntProof {
+    nonce1: Uint8Array;
+    nonce2: Uint8Array;
+}
+
+export interface ILocationData {
+    ent: mcl.G2;
+    pEnt: IEntProof;
+    mtr: IMasterTrace;
+}
+
+export interface IPreTrace {
+    id: Uint8Array;
+    mskhEnc: Uint8Array; // In the paper ctxtHA
+    pskidl: mcl.G1;
+}
+
+export interface ITraceProof {
+    mpk: mcl.G2;
+    nonce1: Uint8Array;
+    nonce2: Uint8Array;
+}
+
+export interface ITrace {
+    id: Uint8Array;
+    skid: mcl.G1;
+}
+
+
+/**
+ * Implements the cryptographic protocol in section 4.2 of the CrowdNotifier whitepaper.
+ */
+export class CrowdNotifierPrimitives {
+
+    /**
+     * Generate key pair for health authority.
+     */
+    static setupHA(): IKeyPair {
+        return crypto_box_keypair();
+    }
+
+
+    /**
+     * Generate an identifier.
+     * @param info public information
+     * @param cnt counter
+     * @param nonce1 a nonce
+     * @param nonce2 an other nonce
+     * @returns an identifier
+     */
+    static genId(info: string, cnt: number, nonce1: Uint8Array, nonce2: Uint8Array): Uint8Array {
```

Comment:
I would not add `genID` here, as it is never defined in the whitepaper. It should rather be a helper method, like I did before, and not part of `CrowdNotifierPrimitives`.